File: test-suite/src/main/java/io/micronaut/gcp/pubsub/serdes/XmlMessageSerDes.java
package io.micronaut.gcp.pubsub.serdes;

import io.micronaut.core.serialize.exceptions.SerializationException;
import io.micronaut.core.type.Argument;
import io.micronaut.http.MediaType;
import jakarta.inject.Named;
import jakarta.inject.Singleton;
import tools.jackson.core.JacksonException;
import tools.jackson.dataformat.xml.XmlMapper;

@Singleton
public class XmlMessageSerDes implements PubSubMessageSerDes {

    private final XmlMapper xmlMapper;

    public XmlMessageSerDes(@Named("xml") XmlMapper xmlMapper) {
        this.xmlMapper = xmlMapper;
    }

    @Override
    public Object deserialize(byte[] data, Argument<?> type) {
        try {
            return xmlMapper.readValue(data, type.getType());
        } catch (JacksonException e) {
            throw new SerializationException("Failed to deserialize PubSub message as XML", e);
        }
    }

    @Override
    public byte[] serialize(Object data) {
        try {
            return xmlMapper.writeValueAsBytes(data);
        } catch (JacksonException e) {
            throw new SerializationException("Failed to serialize PubSub message as XML", e);
        }
    }

    @Override
    public String supportedType() {
        return MediaType.APPLICATION_XML;
    }
}

File: test-suite-python/src/main/python/micronaut/gcp/pubsub/serdes/XmlMessageSerDes.py
from typing import Annotated

from jakarta.inject import Named, Singleton
from micronaut.core.serialize.exceptions import SerializationException
from micronaut.core.type import Argument
from micronaut.gcp.pubsub.serdes import PubSubMessageSerDes
from micronaut.http import MediaType
from tools.jackson.dataformat.xml import XmlMapper


@Singleton
class XmlMessageSerDes(PubSubMessageSerDes):

    def __init__(self, xml_mapper: Annotated[XmlMapper, Named("xml")]):
        self.xml_mapper = xml_mapper

    def deserialize(self, data: bytes, type: Argument) -> object:
        try:
            return self.xml_mapper.readValue(data, type.getType())
        except Exception as e:
            raise SerializationException("Failed to deserialize PubSub message as XML", e)

    def serialize(self, data: object) -> bytes:
        try:
            return self.xml_mapper.writeValueAsBytes(data)
        except Exception as e:
            raise SerializationException("Failed to serialize PubSub message as XML", e)

    def supportedType(self) -> str:
        return MediaType.APPLICATION_XML
